In [49]:
import numpy as np
import pandas as pd

In [125]:
def f(x, y):
    ans = (1.5 - x + x*y)**2 + (2.25 - x + x*y**2)**2 + (2.625 - x + x*y**3)**2
    return ans    

def df_dx(x,y):
    ans = 2 * (1.5 - x + x*y)*(-1+y) + 2 * (2.25 - x + x*y**2)*(-1 + y**2) + 2*(2.625 - x + x*y**3)*(-1 + y**3)
    return ans

def df_dy(x,y):
    ans = 2*(1.5 - x + x*y)*x + 2*(2.25 - x + x*y**2)*2*x*y + 2*(2.625 - x + x*y**3)*3*x*y**2
    return ans

In [153]:
def min_grad(x, y, n_iter=10**3, step=0.01, cond=1e-07):
    # поиск минимума функции методом градиентного спуска
    
    J = f(x,y)        
    for i in range(n_iter):
        
        x -= step * df_dx(x, y)
        y -= step * df_dy(x, y)
        
        if abs(f(x,y)**2 - J**2) <= cond:
        #if np.sqrt(df_dx(x, y)**2 + df_dy(x, y)**2) <= cond:
            print('Задача решена, кол-во итераций = ', i)
            return x, y
        else:
            J = f(x,y)
    
    return 'Минимум не найден'      

def min_grad_moment(x, y, n_iter=10**3, step=0.01, cond=1e-07, gamma=0.9):
    # поиск минимума функции методом моментов
    J = f(x,y)
    U_x , U_y = 0, 0
    for i in range(n_iter):
        
        U_x = gamma*U_x + step* df_dx(x, y)
        U_y = gamma*U_y + step* df_dy(x, y)        
        x -= U_x
        y -= U_y
        
        if abs(f(x,y)**2 - J**2) <= cond:
        #if np.sqrt(df_dx(x, y)**2 + df_dy(x, y)**2) <= cond:
            print('Задача решена, кол-во итераций = ', i)
            return x, y
        else:
            J = f(x,y)
    
    return 'Минимум не найден'     

In [134]:
x0, y0 = 0.7, 1.4
step = 0.01
max_iter = 10**3

#### Метод градиентного спуска

In [138]:
min_grad(x0, y0, n_iter=10**3, step=0.01, cond=1e-07)

Задача решена, кол-во итераций =  425


(2.883538442303829, 0.46951821739369337)

In [136]:
min_grad(3.1, 0.4, n_iter=10**3, step=0.01, cond=1e-07)

Задача решена, кол-во итераций =  7


(3.0642437350915257, 0.5149299414298791)

#### Метод моментов

In [154]:
min_grad_moment(x0, y0, n_iter=10**3, step=0.01, cond=1e-07, gamma=0.9)

Задача решена, кол-во итераций =  304


(3.0723915860096747, 0.5175133259636981)

#### Метод Нестерова

In [173]:
def dU_x_dx(x, y):
    #U_x =  df_dx(x, y)
    #U_x = 2 * (1.5 - x + x*y)*(-1+y) + 2 * (2.25 - x + x*y**2)*(-1 + y**2) + 2*(2.625 - x + x*y**3)*(-1 + y**3)      
    ans = 2 * ( - 1 + y)*(-1+y) + 2 * ( - 1 + y**2)*(-1 + y**2) + 2*(- 1 + y**3)*(-1 + y**3)    
    return ans

def dU_y_dy(x, y):
    #U_y = df_dy(x, y)
    #U_y =  2*(1.5 - x + x*y)*x + 2*(2.25 - x + x*y**2)*2*x*y + 2*(2.625 - x + x*y**3)*3*x*y**2    
    ans = 2*x*x + 2*(2.25*2*x  - x*2*x  + 6*x*x*y**2) + 2*(2.625*6*x*y - x*6*x*y + 15*x*x*y**4)
    return ans



def min_grad_nesterov(x, y, n_iter=10**3, step=0.01, cond=1e-07, gamma=0.9):   
    # поиск минимума функции методом моментов
    J = f(x,y)
    U_x , U_y = 0, 0
    for i in range(n_iter):
        
        #U_x = gamma*U_x + step* df_dx(x, y) - step*gamma*dU_x_dx(x, y)
        #U_y = gamma*U_y + step* df_dy(x, y) - step*gamma*dU_y_dy(x,y)    
        U_x = gamma*U_x + step* df_dx(x, y) - step*gamma*U_x
        U_y = gamma*U_y + step* df_dy(x, y) - step*gamma*U_y
        x -= U_x
        y -= U_y
        
        if abs(f(x,y)**2 - J**2) <= cond:
        #if np.sqrt(df_dx(x, y)**2 + df_dy(x, y)**2) <= cond:
            print('Задача решена, кол-во итераций = ', i)
            return x, y
        else:
            J = f(x,y)
    
    return 'Минимум не найден'     

In [174]:
min_grad_nesterov(x0, y0, n_iter=10**3, step=0.01, cond=1e-07, gamma=0.9)

Задача решена, кол-во итераций =  246


(3.0767817591457387, 0.5185400072747123)

Определим точку, в которой достигается глобальный минимум данной функции

In [186]:
np.linspace(-5,10,25)

array([-5.   , -4.375, -3.75 , -3.125, -2.5  , -1.875, -1.25 , -0.625,
        0.   ,  0.625,  1.25 ,  1.875,  2.5  ,  3.125,  3.75 ,  4.375,
        5.   ,  5.625,  6.25 ,  6.875,  7.5  ,  8.125,  8.75 ,  9.375,
       10.   ])

In [209]:
dot = []
for x in np.linspace(-0.5, 3, 50):
    for y in np.linspace(-0.5, 3, 50):
        dot.append(min_grad_nesterov(x, y))


Задача решена, кол-во итераций =  81
Задача решена, кол-во итераций =  78
Задача решена, кол-во итераций =  81
Задача решена, кол-во итераций =  72
Задача решена, кол-во итераций =  135
Задача решена, кол-во итераций =  142
Задача решена, кол-во итераций =  125
Задача решена, кол-во итераций =  97
Задача решена, кол-во итераций =  138
Задача решена, кол-во итераций =  84
Задача решена, кол-во итераций =  78
Задача решена, кол-во итераций =  91
Задача решена, кол-во итераций =  84
Задача решена, кол-во итераций =  87
Задача решена, кол-во итераций =  80
Задача решена, кол-во итераций =  100
Задача решена, кол-во итераций =  97
Задача решена, кол-во итераций =  106


C:\Users\iziba_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: RuntimeWarning: overflow encountered in double_scalars
  
C:\Users\iziba_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: RuntimeWarning: overflow encountered in double_scalars
  
C:\Users\iziba_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:10: RuntimeWarning: overflow encountered in double_scalars
  # Remove the CWD from sys.path while we load stuff.
C:\Users\iziba_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: RuntimeWarning: invalid value encountered in double_scalars
  
C:\Users\iziba_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: RuntimeWarning: invalid value encountered in double_scalars
  
C:\Users\iziba_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:10: RuntimeWarning: invalid value encountered in double_scalars
  # Remove the CWD from sys.path while we load stuff.


Задача решена, кол-во итераций =  108
Задача решена, кол-во итераций =  92
Задача решена, кол-во итераций =  98
Задача решена, кол-во итераций =  190
Задача решена, кол-во итераций =  211
Задача решена, кол-во итераций =  85
Задача решена, кол-во итераций =  100
Задача решена, кол-во итераций =  719
Задача решена, кол-во итераций =  872
Задача решена, кол-во итераций =  201
Задача решена, кол-во итераций =  221


C:\Users\iziba_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:28: RuntimeWarning: overflow encountered in double_scalars
C:\Users\iziba_000\Anaconda3\lib\site-packages\ipykernel_launcher.py:28: RuntimeWarning: invalid value encountered in double_scalars


Задача решена, кол-во итераций =  331
Задача решена, кол-во итераций =  429
Задача решена, кол-во итераций =  92
Задача решена, кол-во итераций =  177
Задача решена, кол-во итераций =  82
Задача решена, кол-во итераций =  90
Задача решена, кол-во итераций =  94
Задача решена, кол-во итераций =  82
Задача решена, кол-во итераций =  90
Задача решена, кол-во итераций =  93
Задача решена, кол-во итераций =  85
Задача решена, кол-во итераций =  91
Задача решена, кол-во итераций =  152
Задача решена, кол-во итераций =  127
Задача решена, кол-во итераций =  88
Задача решена, кол-во итераций =  93
Задача решена, кол-во итераций =  85
Задача решена, кол-во итераций =  81
Задача решена, кол-во итераций =  92
Задача решена, кол-во итераций =  94
Задача решена, кол-во итераций =  92
Задача решена, кол-во итераций =  109
Задача решена, кол-во итераций =  423
Задача решена, кол-во итераций =  104
Задача решена, кол-во итераций =  90
Задача решена, кол-во итераций =  90
Задача решена, кол-во итераций

Задача решена, кол-во итераций =  107
Задача решена, кол-во итераций =  102
Задача решена, кол-во итераций =  116
Задача решена, кол-во итераций =  357
Задача решена, кол-во итераций =  104
Задача решена, кол-во итераций =  106
Задача решена, кол-во итераций =  172
Задача решена, кол-во итераций =  110
Задача решена, кол-во итераций =  97
Задача решена, кол-во итераций =  94
Задача решена, кол-во итераций =  102
Задача решена, кол-во итераций =  715
Задача решена, кол-во итераций =  108
Задача решена, кол-во итераций =  83
Задача решена, кол-во итераций =  89
Задача решена, кол-во итераций =  77
Задача решена, кол-во итераций =  78
Задача решена, кол-во итераций =  64
Задача решена, кол-во итераций =  84
Задача решена, кол-во итераций =  67
Задача решена, кол-во итераций =  92
Задача решена, кол-во итераций =  82
Задача решена, кол-во итераций =  69
Задача решена, кол-во итераций =  78
Задача решена, кол-во итераций =  89
Задача решена, кол-во итераций =  78
Задача решена, кол-во итера

Задача решена, кол-во итераций =  70
Задача решена, кол-во итераций =  81
Задача решена, кол-во итераций =  83
Задача решена, кол-во итераций =  85
Задача решена, кол-во итераций =  77
Задача решена, кол-во итераций =  88
Задача решена, кол-во итераций =  89
Задача решена, кол-во итераций =  74
Задача решена, кол-во итераций =  77
Задача решена, кол-во итераций =  65
Задача решена, кол-во итераций =  84
Задача решена, кол-во итераций =  83
Задача решена, кол-во итераций =  79
Задача решена, кол-во итераций =  84
Задача решена, кол-во итераций =  74
Задача решена, кол-во итераций =  87
Задача решена, кол-во итераций =  85
Задача решена, кол-во итераций =  88
Задача решена, кол-во итераций =  84
Задача решена, кол-во итераций =  89
Задача решена, кол-во итераций =  80
Задача решена, кол-во итераций =  71
Задача решена, кол-во итераций =  89
Задача решена, кол-во итераций =  94
Задача решена, кол-во итераций =  94
Задача решена, кол-во итераций =  85
Задача решена, кол-во итераций =  94
З

Задача решена, кол-во итераций =  94
Задача решена, кол-во итераций =  91
Задача решена, кол-во итераций =  259
Задача решена, кол-во итераций =  83
Задача решена, кол-во итераций =  107
Задача решена, кол-во итераций =  119
Задача решена, кол-во итераций =  119
Задача решена, кол-во итераций =  119
Задача решена, кол-во итераций =  119
Задача решена, кол-во итераций =  117
Задача решена, кол-во итераций =  113
Задача решена, кол-во итераций =  69
Задача решена, кол-во итераций =  79
Задача решена, кол-во итераций =  79
Задача решена, кол-во итераций =  72
Задача решена, кол-во итераций =  71
Задача решена, кол-во итераций =  65
Задача решена, кол-во итераций =  72
Задача решена, кол-во итераций =  83
Задача решена, кол-во итераций =  45
Задача решена, кол-во итераций =  86
Задача решена, кол-во итераций =  82
Задача решена, кол-во итераций =  87
Задача решена, кол-во итераций =  87
Задача решена, кол-во итераций =  87
Задача решена, кол-во итераций =  87
Задача решена, кол-во итераций

Задача решена, кол-во итераций =  123
Задача решена, кол-во итераций =  71
Задача решена, кол-во итераций =  70
Задача решена, кол-во итераций =  61
Задача решена, кол-во итераций =  66
Задача решена, кол-во итераций =  69
Задача решена, кол-во итераций =  61
Задача решена, кол-во итераций =  49
Задача решена, кол-во итераций =  20
Задача решена, кол-во итераций =  44
Задача решена, кол-во итераций =  61
Задача решена, кол-во итераций =  62
Задача решена, кол-во итераций =  55
Задача решена, кол-во итераций =  61
Задача решена, кол-во итераций =  67
Задача решена, кол-во итераций =  73
Задача решена, кол-во итераций =  70
Задача решена, кол-во итераций =  71
Задача решена, кол-во итераций =  76
Задача решена, кол-во итераций =  85
Задача решена, кол-во итераций =  90
Задача решена, кол-во итераций =  88
Задача решена, кол-во итераций =  87
Задача решена, кол-во итераций =  88
Задача решена, кол-во итераций =  99
Задача решена, кол-во итераций =  99
Задача решена, кол-во итераций =  60


Задача решена, кол-во итераций =  85
Задача решена, кол-во итераций =  76
Задача решена, кол-во итераций =  71
Задача решена, кол-во итераций =  75
Задача решена, кол-во итераций =  70
Задача решена, кол-во итераций =  56
Задача решена, кол-во итераций =  73
Задача решена, кол-во итераций =  68
Задача решена, кол-во итераций =  62
Задача решена, кол-во итераций =  57
Задача решена, кол-во итераций =  47
Задача решена, кол-во итераций =  42
Задача решена, кол-во итераций =  38
Задача решена, кол-во итераций =  51
Задача решена, кол-во итераций =  61
Задача решена, кол-во итераций =  67
Задача решена, кол-во итераций =  73
Задача решена, кол-во итераций =  70
Задача решена, кол-во итераций =  75
Задача решена, кол-во итераций =  85
Задача решена, кол-во итераций =  91
Задача решена, кол-во итераций =  101
Задача решена, кол-во итераций =  90
Задача решена, кол-во итераций =  115
Задача решена, кол-во итераций =  85
Задача решена, кол-во итераций =  76
Задача решена, кол-во итераций =  80

In [210]:
all_x = []
all_y = []
for xy in dot:
    if len(xy) == 2:
        all_x.append(xy[0])
        all_y.append(xy[1])
        
print('Точка, в которой достигается минимум функции имеет координаты (', round(np.mean(all_x), 2), ', ', round(np.mean(all_y),2), ').')
        

Точка, в которой достигается минимум функции имеет координаты ( 2.99 ,  0.5 ).


### Вывод
<br> 1. Cогласно построенной модели фунция достигает минимума в точке `(2.99, 0.5)`. Согласно WolframMatematica функция достигает своего экстремума в точке (3, 0.5). Результаты хорошо согласованы, погрешность меньше 1%. <br> 2. Реализованные методы сходятся за различное число итераций. Наихудший результат показал обычный `градиентный спуск` (*425 итераций*). Лучше справился `метод моментов` (*304 итерации*). Непревзойдённым победителем стал `метод Нестерова` (*246 итераций*)
